File: packages/sdk-flutter/lib/src/hmac.dart
import 'dart:convert';
import 'package:crypto/crypto.dart';

import 'types.dart';

/// Canonical string to HMAC-sign, matching apps/server/src/hmac.ts:
///   "<METHOD>\n<path>\n<timestamp>\n<nonce>\n<body>"
String canonicalString(
  String method,
  String path,
  String timestamp,
  String nonce,
  String body,
) {
  return [method.toUpperCase(), path, timestamp, nonce, body].join('\n');
}

/// Lower-case hex HMAC-SHA256.
String signHmac(String secret, String data) {
  final mac = Hmac(sha256, utf8.encode(secret));
  return mac.convert(utf8.encode(data)).toString();
}

/// Fields included in the hostContext signature, in the exact order the
/// server canonicalizes them. Must mirror `CANONICAL_FIELDS` in
/// packages/core/src/hostContext.ts.
const _canonicalHostContextFields = <String>[
  'uid',
  'cookieHash',
  'sessionHash',
  'accountAgeDays',
  'emailDomainHash',
  'kycLevel',
  'tags',
  'verifiedIdentities',
];

/// Canonical JSON for HMAC-signing a [HostContext]. Mirrors
/// `canonicalizeHostContext` in packages/core/src/hostContext.ts: a JSON
/// array of `[key, value]` pairs in [_canonicalHostContextFields] order,
/// with array values lexicographically sorted.
String _canonicalizeHostContext(HostContext ctx) {
  final values = <String, dynamic>{};
  if (ctx.uid != null) values['uid'] = ctx.uid;
  if (ctx.cookieHash != null) values['cookieHash'] = ctx.cookieHash;
  if (ctx.sessionHash != null) values['sessionHash'] = ctx.sessionHash;
  if (ctx.accountAgeDays != null) values['accountAgeDays'] = ctx.accountAgeDays;
  if (ctx.emailDomainHash != null) {
    values['emailDomainHash'] = ctx.emailDomainHash;
  }
  if (ctx.kycLevel != null) values['kycLevel'] = ctx.kycLevel;
  if (ctx.tags != null && ctx.tags!.isNotEmpty) {
    final sorted = List<String>.from(ctx.tags!)..sort();
    values['tags'] = sorted;
  }
  // verifiedIdentities not yet on the Dart HostContext; the canonical
  // order below already accounts for it so once added it lines up.
  final entries = <List<dynamic>>[];
  for (final k in _canonicalHostContextFields) {
    final v = values[k];
    if (v == null) continue;
    entries.add([k, v]);
  }
  return jsonEncode(entries);
}

/// Sign a [HostContext] with an integrator HMAC secret and return a copy
/// whose `signature` is set to `<integratorId>:<base64-hmac>`. Mirrors
/// `FaucetClient.signHostContext` in packages/sdk-ts/src/index.ts and
/// `SignHostContext` in packages/sdk-go/hmac.go (closes audit
/// Improvement #104).
///
/// Run this on your BACKEND — never expose [hmacSecret] to the device.
/// Pass the signed context through to the on-device SDK's claim() call.
HostContext signHostContext(
  HostContext ctx,
  String integratorId,
  String hmacSecret,
) {
  final canonical = _canonicalizeHostContext(ctx);
  final mac = Hmac(sha256, utf8.encode(hmacSecret));
  final digest = mac.convert(utf8.encode(canonical));
  final sig = '$integratorId:${base64.encode(digest.bytes)}';
  return HostContext(
    uid: ctx.uid,
    cookieHash: ctx.cookieHash,
    sessionHash: ctx.sessionHash,
    accountAgeDays: ctx.accountAgeDays,
    emailDomainHash: ctx.emailDomainHash,
    kycLevel: ctx.kycLevel,
    tags: ctx.tags,
    signature: sig,
  );
}

File: packages/sdk-go/hmac.go
package faucet

import (
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/base64"
	"encoding/hex"
	"encoding/json"
	"sort"
	"strings"
)

// canonicalString assembles the string the server HMAC-signs:
//
//	"POST\n<path>\n<timestamp>\n<nonce>\n<body>"
//
// Must match apps/server/src/hmac.ts#canonicalString exactly.
func canonicalString(method, path, timestamp, nonce, body string) string {
	return strings.Join([]string{strings.ToUpper(method), path, timestamp, nonce, body}, "\n")
}

// signHMAC returns the lower-case hex HMAC-SHA256 of `data` using `secret`.
func signHMAC(secret, data string) string {
	mac := hmac.New(sha256.New, []byte(secret))
	mac.Write([]byte(data))
	return hex.EncodeToString(mac.Sum(nil))
}

// randomNonce returns a 16-byte random nonce as lowercase hex. Falls back to
// an empty string on entropy failure; callers should treat empty as a signal
// to retry (extraordinarily unlikely on any supported platform).
func randomNonce() string {
	buf := make([]byte, 16)
	if _, err := rand.Read(buf); err != nil {
		return ""
	}
	return hex.EncodeToString(buf)
}

// canonicalHostContextFields lists the fields included in the
// hostContext signature, in the exact order the server expects. Must
// match `CANONICAL_FIELDS` in packages/core/src/hostContext.ts.
var canonicalHostContextFields = [...]string{
	"uid",
	"cookieHash",
	"sessionHash",
	"accountAgeDays",
	"emailDomainHash",
	"kycLevel",
	"tags",
	"verifiedIdentities",
}

// canonicalizeHostContext serialises the trust-claim fields of a HostContext
// in the same shape the server canonicalizes for HMAC signing — a JSON
// array of [key, value] pairs in CANONICAL_FIELDS order, with array
// values lexicographically sorted. Mirrors
// packages/core/src/hostContext.ts#canonicalizeHostContext.
//
// Note: HostContext currently has no VerifiedIdentities field on the
// Go struct (the server schema does); when that field is added, plumb
// it through here too.
func canonicalizeHostContext(ctx *HostContext) (string, error) {
	if ctx == nil {
		return "[]", nil
	}
	type entry struct {
		key   string
		value interface{}
	}
	values := map[string]interface{}{}
	if ctx.UID != nil {
		values["uid"] = *ctx.UID
	}
	if ctx.CookieHash != nil {
		values["cookieHash"] = *ctx.CookieHash
	}
	if ctx.SessionHash != nil {
		values["sessionHash"] = *ctx.SessionHash
	}
	if ctx.AccountAgeDays != nil {
		values["accountAgeDays"] = *ctx.AccountAgeDays
	}
	if ctx.EmailDomainHash != nil {
		values["emailDomainHash"] = *ctx.EmailDomainHash
	}
	if ctx.KYCLevel != nil {
		values["kycLevel"] = *ctx.KYCLevel
	}
	if len(ctx.Tags) > 0 {
		// Sort a copy so we don't mutate the caller's slice.
		sorted := append([]string(nil), ctx.Tags...)
		sort.Strings(sorted)
		values["tags"] = sorted
	}
	// verifiedIdentities not yet on the Go HostContext struct; leave a
	// hook here so the canonical order matches the server when it is.
	entries := make([][2]interface{}, 0, len(canonicalHostContextFields))
	for _, k := range canonicalHostContextFields {
		v, ok := values[k]
		if !ok {
			continue
		}
		entries = append(entries, [2]interface{}{k, v})
	}
	out, err := json.Marshal(entries)
	if err != nil {
		return "", err
	}
	return string(out), nil
}

// SignHostContext signs a HostContext with an integrator HMAC secret and
// returns a copy whose Signature is set to "<integratorID>:<base64-hmac>".
// The signature uses HMAC-SHA256 over the canonical JSON the server
// reconstructs; mirrors `FaucetClient.signHostContext` in the TS SDK
// (packages/sdk-ts/src/index.ts).
//
// Run this on your BACKEND — never expose hmacSecret to the browser.
// Pass the signed context through to the browser SDK's Claim() call
// (closes audit Improvement #104).
func SignHostContext(ctx HostContext, integratorID, hmacSecret string) (HostContext, error) {
	canonical, err := canonicalizeHostContext(&ctx)
	if err != nil {
		return HostContext{}, err
	}
	mac := hmac.New(sha256.New, []byte(hmacSecret))
	mac.Write([]byte(canonical))
	sig := integratorID + ":" + base64.StdEncoding.EncodeToString(mac.Sum(nil))
	out := ctx
	out.Signature = &sig
	return out, nil
}
